File: src/ski_area.rs
use geo::Point;
use serde::{Deserialize, Serialize};

use lift::parse_lift;
use piste::parse_pistes;

use crate::config::get_config;
use crate::error::{Error, ErrorType, Result};
use crate::osm_reader::{get_tag, Document};

mod bounded_geometry;
mod lift;
mod piste;

#[cfg(test)]
mod lift_test;
#[cfg(test)]
mod piste_test;

pub use bounded_geometry::BoundedGeometry;
pub use lift::Lift;
pub use piste::{Difficulty, Piste, PisteData, PisteMetadata};

#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
pub struct PointWithElevation {
    pub point: Point,
    pub elevation: u32,
}

impl PointWithElevation {
    pub fn new(point: Point, elevation: u32) -> Self {
        Self { point, elevation }
    }
}

#[derive(Serialize, Deserialize, Debug)]
pub struct SkiArea {
    pub name: String,
    pub lifts: Vec<Lift>,
    pub pistes: Vec<Piste>,
}

impl SkiArea {
    pub fn parse(doc: &Document) -> Result<Self> {
        let mut names: Vec<String> = Vec::new();
        let mut lifts = Vec::new();
        let config = get_config();

        for (_id, way) in &doc.elements.ways {
            if get_tag(&way.tags, "landuse") == "winter_sports" {
                names.push(get_tag(&way.tags, "name").to_string());
            }
        }

        if names.len() == 0 {
            return Err(Error::new_s(
                ErrorType::InputError,
                "ski area entity not found",
            ));
        } else if names.len() > 1 {
            return Err(Error::new(
                ErrorType::InputError,
                format!("ambiguous ski area: {:?}", names),
            ));
        }

        for (id, way) in &doc.elements.ways {
            match parse_lift(&doc, &id, &way) {
                Err(e) => eprintln!("Error parsing way {}: {}", id, e),
                Ok(None) => (),
                Ok(Some(lift)) => lifts.push(lift),
            }
        }

        if config.is_v() {
            eprintln!("Found {} lifts.", lifts.len());
        }

        let pistes = parse_pistes(&doc);

        if config.is_v() {
            eprintln!("Found {} pistes.", pistes.len());
        }

        Ok(SkiArea {
            name: names.remove(0),
            lifts,
            pistes,
        })
    }
}

File: src/app/types/skiArea.ts
import { Rect } from "./geo";
import type { Lift, Piste, SkiAreaMetadata } from "./generated/generated";
import { indexData } from "@/utils/data";
import { SkiArea as RawSkiArea } from "./generated/generated";

export {
  PointWithElevation,
  Difficulty,
  Lift,
  PisteMetadata,
  PisteData,
  Piste,
  SkiAreaMetadata,
} from "./generated/generated";

export { RawSkiArea };

export type SkiArea = {
  metadata: SkiAreaMetadata;
  lifts: Map<string, Lift>;
  pistes: Map<string, Piste>;
  bounding_rect: Rect;
  date: string;
};

export function indexSkiArea(ski_area: RawSkiArea): SkiArea {
  return {
    metadata: ski_area.metadata,
    lifts: indexData<Lift>(ski_area.lifts),
    pistes: indexData<Piste>(ski_area.pistes),
    bounding_rect: ski_area.bounding_rect,
    date: ski_area.date,
  };
}
